File: CTCI/JavaScript/01_Arrays_and_Strings/08_zero_matrix.js
/**
 * Problem:
 * Write an algorithm such that if an element in an MxN matrix is 0, its entire
 * row and column are set to zero.
 * 
 * M = X dimension of matrix
 * N = Y dimension of matrix
 * 
 * Hints:
 * - If you just cleared the rows and columns as you found 0s, you'd likely
 *   wind up clearing the whole matrix. Try finding the cells with zeros first
 *   before making any changes to the matrix.
 * - Can you use O(N) additional space instead of O(N^2)? What information do
 *   you really need from the list of cells that are zero?
 * - You probably need some data storage to maintain a list of the rows and
 *   columns that need to be zeroed. Can you reduce the additional space usage
 *   to O(1) by using the matrix itself for data storage?
 * 
 * [1, 2, 3]
 * [1, 0, 3]
 * [1, 2, 3]
 * 
 * @param  {number[][]} matrix Matrix to be zeroed
 * @return {number[][]}        Zeroed matrix
 */

function zeroMatrix(matrix) {
  if (!matrix) throw new Error('invalid matrix');
  if (matrix.length === 0) return matrix;

  let R = matrix.length;
  let C = matrix[0].length;
  let colHasZero = false;

  for (let i = 0; i < R; i++) {
    if (matrix[i][0] === 0) {
      colHasZero = true;
    }

    // if an element is 0, set 1st element of corresponding row & column to 0
    for (let j = 1; j < C; j++) {
      if (matrix[i][j] === 0) {
        matrix[0][j] = 0;
        matrix[i][0] = 0;
      }
    }
  }

  // iterate over matrix again and using the 1st row and column, update elements
  for (let i = 1; i < R; i++) {
    for (let j = 1; j < C; j++) {
      if (matrix[i][0] === 0 || matrix[0][j] === 0) {
        matrix[i][j] = 0;
      }
    }
  }

  // check if the first row needs to be zeroed
  if (matrix[0][0] === 0) {
    for (let j = 0; j < C; j++) {
      matrix[0][j] = 0;
    }
  }

  // check if the first column needs to be zeroed
  if (colHasZero) {
    for (let i = 0; i < R; i++) {
      matrix[i][0] = 0;
    }
  }
  return matrix;
}

File: CTCI/Java/01_Arrays_and_Strings/08_zero_matrix.java
/**
 * Problem:
 * Write an algorithm such that if an element in an MxN matrix is 0, its entire
 * row and column are set to zero.
 * 
 * Example 1:
 * Input: 
 * [
 *   [1,1,1],
 *   [1,0,1],
 *   [1,1,1]
 * ]
 * Output: 
 * [
 *   [1,0,1],
 *   [0,0,0],
 *   [1,0,1]
 * ]
 * 
 * Example 2:
 * Input: 
 * [
 *   [0,1,2,0],
 *   [3,4,5,2],
 *   [1,3,1,5]
 * ]
 * Output: 
 * [
 *   [0,0,0,0],
 *   [0,4,5,0],
 *   [0,3,1,0]
 * ]
 * 
 * @param matrix Matrix to be zeroed
 * @return The zeroed matrix
 */

class ZeroMatrix {
    void zeroMatrix(int[][] matrix) {
        boolean rowHasZero = false;
        boolean colHasZero = false;

        // check if the first row contains a zero
        for (int col = 0; col < matrix[0].length; col++) {
            if (matrix[0][col] == 0) {
                rowHasZero = true;
                break;
            }
        }

        // check if the first column contains a zero
        for (int row = 0; row < matrix.length; row++) {
            if (matrix[row][0] == 0) {
                colHasZero = true;
                break;
            }
        }

        // check for zeroes in the rest of the array
        for (int row = 1; row < matrix.length; row++) {
            for (int col = 1; col < matrix[0].length; col++) {
                if (matrix[row][col] == 0) {
                    matrix[row][0] = 0;
                    matrix[0][col] = 0;
                }
            }
        }

        // zero the rows based on the values in the first column
        for (int row = 1; row < matrix.length; row++) {
            if (matrix[row][0] == 0) {
                zeroRow(matrix, row);
            }
        }

        // zero the columns based on the values in the first row
        for (int col = 1; col < matrix[0].length; col++) {
            if (matrix[0][col] == 0) {
                zeroColumn(matrix, col);
            }
        }

        // zero the first row
        if (rowHasZero) {
            zeroRow(matrix, 0);
        }

        // zero the first column
        if (colHasZero) {
            zeroColumn(matrix, 0);
        }
    }

    /* Sets the elements in the row to 0 */
    void zeroRow(int[][] matrix, int row) {
        for (int col = 0; col < matrix[0].length; col++) {
            matrix[row][col] = 0;
        }
    }

    /* Sets the elements in the column to 0 */
    void zeroColumn(int[][] matrix, int col) {
        for (int row = 0; row < matrix[0].length; row++) {
            matrix[row][col] = 0;
        }
    }
}